subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/GroovyJavaJointCompileSourceOrderIntegrationTest.groovy

```diff
@@ -54,6 +55,36 @@ class GroovyJavaJointCompileSourceOrderIntegrationTest extends AbstractIntegrati
         reversedAgainBytes == originalBytes
     }
 
+    @Unroll
+    def "groovy and java source directory compilation order can be reversed (task configuration #configurationStyle)"() {
+        given:
+        file("src/main/groovy/Groovy.groovy") << "class Groovy { }"
+        file("src/main/java/Java.java") << "public class Java { Groovy groovy = new Groovy(); }"
```

Comment:
Stepping back, this test case should really fail, because there's no explicit dependency declared between from the Java to Groovy source set. The convention should really only be that the Groovy source set depends on the Java source set, which the test discards. I know this isn't what we're trying to fix here, but something we should consider addressing at some point.